File: src/ch15/Mixins.java
/**
 * @Title Mixins.java
 * @Package ch15
 * @Description TODO
 * @author 吴扬颉
 * @date 2019年3月28日
 * @version 1.0
 */
package ch15;
import java.sql.Time;
import java.util.*;
interface TimeStamped { long getStamp(); }
class TimeStampedImp implements TimeStamped{
	private final long timeStamp;
	public TimeStampedImp(){
		timeStamp = new Date().getTime();
	}
	public long getStamp(){ return timeStamp; }
}

interface SerialNumbered{ long getSerialNumber(); }
class SerialNumberedImp implements SerialNumbered{
	private static long counter = 1;
	private final long serialNumer = counter++;
	public long getSerialNumber(){ return serialNumer; }
}

interface Basic{
	public void set(String val);
	public String get();
}

class BasicImp implements Basic{
	private String value;
	public void set(String val){ value = val; }
	public String get(){ return value; }
}
// 该类基本上是在使用代理，因此每个混入类型都要求在Mixin中有一个对应的域。
// 而且必须在Mixin中编写所有必须的方法，将方法调用转发给恰当的对象
class Mixin extends BasicImp implements TimeStamped, SerialNumbered{
	private TimeStamped timeStamp = new TimeStampedImp();
	private SerialNumbered serialNumber = new SerialNumberedImp();
	public long getStamp(){ return timeStamp.getStamp(); }
	public long getSerialNumber(){ return serialNumber.getSerialNumber(); }
}

/**
 * @ClassName Mixins
 * @Description 由于擦除的存在，Java不能像C++那样使用混型。一种推荐方案是使用接口来产生混型效果。
 * @author 吴扬颉
 * @date 2019年3月28日
 * 
 */
public class Mixins {
	public static void main(String[] args) {
		Mixin mixin1 = new Mixin(), mixin2 = new Mixin();
		mixin1.set("test string 1");
		mixin2.set("test string 2");
		System.out.println(mixin1.get() + " " + mixin1.getStamp() + " " + mixin1.getSerialNumber());
		System.out.println(mixin2.get() + " " + mixin2.getStamp() + " " + mixin2.getSerialNumber());
	}
}

File: src/ch15/Mixins.cpp
/*
C++使用多重继承的最大理由，就是为了使用混型。但是对于混型来说，更好的是用参数化类型，因为混型就是继承自其它类型参数的类。
在C++可以很容易的创建混型，因为C++能够记住其模板参数的类型
*/
#include <string>
#include <ctime>
#include <iostream>
using namespace std;

template<class T> class TimeStamped : public T {
  long timeStamp;
public:
  TimeStamped() { timeStamp = time(0); }
  long getStamp() { return timeStamp; }
};

template<class T> class SerialNumbered : public T {
  long serialNumber;
  static long counter;
public:
  SerialNumbered() { serialNumber = counter++; }
  long getSerialNumber() { return serialNumber; }
};

// Define and initialize the static storage:
template<class T> long SerialNumbered<T>::counter = 1;

class Basic {
  string value;
public:
  void set(string val) { value = val; }
  string get() { return value; }
};	

int main() {
  TimeStamped<SerialNumbered<Basic> > mixin1, mixin2;
  mixin1.set("test string 1");
  mixin2.set("test string 2");
  cout << mixin1.get() << " " << mixin1.getStamp() <<
    " " << mixin1.getSerialNumber() << endl;
  cout << mixin2.get() << " " << mixin2.getStamp() <<
    " " << mixin2.getSerialNumber() << endl;
} /* Output: (Sample)
test string 1 1129840250 1
test string 2 1129840250 2
*///:~